File: nodejs/src/explorer.ts
import { NETWORK_EXPLORERS } from "./constants.js"
import type { Network } from "./types.js"
import { NETWORK } from "./types.js"

function baseFor(network: Network, override?: string): string {
  const base = override ?? NETWORK_EXPLORERS[network]
  if (!base) throw new Error(`no explorer URL configured for network "${network}"`)
  return base.replace(/\/+$/, "")
}

/**
 * Returns the explorer URL for a transaction hash on `network` (default: BASE).
 *
 * @example
 * txUrl("0xabc...")                       // https://basescan.org/tx/0xabc...
 * txUrl("0xabc...", NETWORK.BSC)          // https://bscscan.com/tx/0xabc...
 * txUrl(hash, NETWORK.BASE, "https://...")// custom explorer base
 */
export function txUrl(hash: string, network: Network = NETWORK.BASE, explorer?: string): string {
  return `${baseFor(network, explorer)}/tx/${hash}`
}

/**
 * Returns the explorer URL for an address on `network` (default: BASE).
 */
export function addressUrl(address: string, network: Network = NETWORK.BASE, explorer?: string): string {
  return `${baseFor(network, explorer)}/address/${address}`
}

File: go/explorer.go
package afi

import (
	"fmt"
	"strings"
)

func explorerBase(network Network, override ...string) (string, error) {
	if len(override) > 0 && override[0] != "" {
		return strings.TrimRight(override[0], "/"), nil
	}
	base, ok := NetworkExplorers[network]
	if !ok {
		return "", fmt.Errorf("no explorer URL configured for network %q", network)
	}
	return strings.TrimRight(base, "/"), nil
}

// TxURL returns the explorer URL for a transaction hash on `network`.
// Pass an explicit `explorer` base to override the default in NetworkExplorers.
//
//	afi.TxURL("0xabc...", afi.NetworkBase)                 // https://basescan.org/tx/0xabc...
//	afi.TxURL(hash, afi.NetworkBase, "https://customexp")  // custom base
func TxURL(hash string, network Network, explorer ...string) (string, error) {
	base, err := explorerBase(network, explorer...)
	if err != nil {
		return "", err
	}
	return base + "/tx/" + hash, nil
}

// AddressURL returns the explorer URL for an address on `network`.
func AddressURL(address string, network Network, explorer ...string) (string, error) {
	base, err := explorerBase(network, explorer...)
	if err != nil {
		return "", err
	}
	return base + "/address/" + address, nil
}
